### PR TITLE
Replace PlotArray with plain DataArray and add redraw() method

### PR DESCRIPTION
**Pull request by nvaytet**
- Remove `PlotArray` in favour of a plain `DataArray` after implementation of shared ownership
- Add `redraw()` function on plots, which update the data in plots after data was updated, without changing axes or color bar.

In `ResamplingModel`, `.data` doesn't need to be a property since it has no setter. We make it into a getter function (`get_data_array`) which then allows to have function arguments, to decide whether we want to skip caching (which is required to update the image even when the axis limits have not changed).

### CODE REVIEW

**Comment by nvaytet** [comment]
@arm61 please try this out and let me know if this is the behaviour you were after.

**Comment by nvaytet** [comment]
Should I add more options to `redraw` (disabled by default) which would allow automatic colorbar rescaling or automatic axis adjustment?

**Comment by nvaytet** [comment]
hmmm just found that it does not update the plot if the coords has been changed instead of the data. e.g.
```Py
N = 50
d = sc.DataArray(
    data=sc.Variable(['tof'], values=5.0 + 5.0*np.sin(np.arange(N, dtype=np.float64)/5.0),
                          unit=sc.units.counts),
    coords={'tof': sc.Variable(['tof'], values=np.arange(N).astype(np.float64),
                                unit=sc.units.us)})
p = sc.plot(d)
p
d.coords['tof'] *= 2
p.redraw()
```

File: python/src/scipp/plotting/resampling_model.py [hunk 1]
```diff
@@ -126,16 +130,13 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # TODO See #1469. This is a temporary hack to work around the
         # conversion of coords to edges in model.py.
-        new_meta = {name: var for name, var in self._array.meta.items()}
-        self._array = PlotArray(data=self._array.data,
-                                masks=self._array.masks,
-                                meta=new_meta)
-        for name, var in self._array.meta.items():
+        self._array = self._array.copy()
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:133]: Why deep copy?

File: python/src/scipp/plotting/resampling_model.py [hunk 2]
```diff
@@ -180,16 +181,20 @@ def __init__(self, array, **kwargs):
...
-                          masks=array.masks)
-        for dim in array.data.dims:
-            coord = array.meta[dim]
+        if array.dtype != sc.dtype.float64:
+            data = array.data.astype(sc.dtype.float64)
+        else:
+            data = array.data
+        density = sc.DataArray(data=data,
+                               coords=array.coords.to_dict(),
+                               masks=array.masks.to_dict())
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:190]:
```suggestion
        array = sc.astype(array, sc.dtype.float64)
```
and keep calling it `array` below.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Haven't tried the redraw functionality yet, here is a look at the code:

**Comment by matthew-d-jones** [comment]
`redraw()` is not updating my plot when used in an async loop, the plot updates only once the loop terminates. Experimenting with a matplotlib plot I see that `draw()` does work in this scenario, but `draw_idle()` does not. Even if I force the event loop to be idle by immediately awaiting an async sleep after calling `draw_idle()` it does not update the plot until the loop terminates.

File: python/src/scipp/plotting/objects.py [hunk 3]
```diff
@@ -6,7 +6,61 @@
...
-class SciPlot:
+class PlotDict(dict):
+    """
+    The Plot object is used as output for the plot command.
+    It is a small wrapper around python dict, with an `_ipython_display_`
+    representation.
+    The dict will contain one entry for each entry in the input supplied to
+    the plot function.
+    More functionalities can be added in the future.
+    """
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plotting/objects.py:17]: We already had that class before, but i'm happy to change it as you suggest. :+1:

File: python/src/scipp/plotting/resampling_model.py [hunk 4]
```diff
@@ -35,11 +34,13 @@ def bounds(self):
     def bounds(self, bnds):
         self._bounds = bnds
 
-    @property
-    def data(self):
-        self._call_resample()
+    def get_data_array(self, force_update=False):
+        self._call_resample(force_update=force_update)
         return self._resampled
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:39]: This was what your `.data` property access was doing.
I just changed it into a function, so I could add an argument to sometimes force the update by not using the cached view.
I can change the name to not have 'get'.

File: python/src/scipp/plotting/resampling_model.py [hunk 5]
```diff
@@ -96,23 +97,26 @@ def _call_resample(self):
...
-                                                      low, high)]
+                    if len(out.coords[dim].dims) == 1:
+                        # TODO: for now, slicing in scipp does not support step
+                        sl = sc.get_slice_params(out.data, out.coords[dim],
+                                                 low, high)
+                        out = out[sl[0], slice(sl[1].start, sl[1].stop)]
                 params[dim] = (low.value, high.value, low.unit,
                                self.resolution[dim])
-        if params == self._home_params:
+        if (params == self._home_params) and (not force_update):
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:111]: I'll give that a go :+1:

File: python/src/scipp/plotting/resampling_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:39]: My concern was not about the name, but the interface in general. Is there a nicer way to trigger updates, e.g., with a new property?

File: python/src/scipp/plotting/resampling_model.py (hunk 1 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:133]: I wanted to shallow copy, and i was thinking of python dicts, and forgot that we decided to deep copy by default.

File: python/src/scipp/plotting/resampling_model.py (hunk 2 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:190]: hmm yes, I had started with that but then remembered that `astype` always makes a new object, even if the dtype is the same.
But you are implying this is ok because even if `astype` is making a copy, it is cheap because of the shared ownership?

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:190]: Yes, it is copying the object but not the buffers.

File: python/src/scipp/plotting/resampling_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:39]: If i do what you suggest below, with the `self._home_params = None` i might not need this change and can go back to `.data`?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:39]: 👍

File: python/src/scipp/plotting/resampling_model.py [hunk 6]
```diff
@@ -96,23 +97,26 @@ def _call_resample(self):
...
                 else:
                     # cannot pre-select bins for multi-dim coords
-                    if len(out.meta[dim].dims) == 1:
-                        out = out[sc.get_slice_params(out.data, out.meta[dim],
-                                                      low, high)]
+                    if len(out.coords[dim].dims) == 1:
+                        # TODO: for now, slicing in scipp does not support step
+                        sl = sc.get_slice_params(out.data, out.coords[dim],
+                                                 low, high)
+                        out = out[sl[0], slice(sl[1].start, sl[1].stop)]
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:106]: This is necessary because `get_slice_params()` always returns a slice with a step of 1, because `py::slice` wants the step to be specified: `py::slice(start, stop, 1)`.
It won't compile with `py::slice(start, stop)`.

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:106]: Are you saying that `__getitem__` fails if you specify a step, even if it is 1?

**Comment by nvaytet** [comment]
@matthew-d-jones 
Please have another look.
- It should now also work with integers.
- Using `draw()` instead of `draw_idle()` made things significantly slower for me. Maybe we can add a flag that you can enable to force it to use `draw()`? e.g. `p.redraw(idle=False)` ? (or a better name?)

**Comment by matthew-d-jones** [comment]
@nvaytet Thanks, I'll give it another go.

> made things significantly slower for me

It will return from `draw()` slower because that causes it to do more of the work during the function call, but it isn't like `draw_idle()` reduces the work or moves the work to a separate thread, so I doubt it would affect total runtime of a workflow or anything. But `draw(idle=False)` sounds fine to me.

**Comment by matthew-d-jones** [comment]
I can confirm it works with integers now. :+1:

**Comment by nvaytet** [comment]
@matthew-d-jones please have a go at the `redraw` inside a loop after setting `p.set_draw_no_delay(True)`.

**Comment by matthew-d-jones** [comment]
@nvaytet The plot now redraws nicely, but any interaction with the plot such as zooming or resizing still does not take effect until data streaming ends.

**Comment by nvaytet** [comment]
> @nvaytet The plot now redraws nicely, but any interaction with the plot such as zooming or resizing still does not take effect until data streaming ends.

Hmmm not sure why that is, if `draw_no_delay` has been set to `True` all updates should behave the same...
Have you tried doing the same (e.g. zooming) on a raw matplotlib figure? Does the same happen there?

**Comment by matthew-d-jones** [comment]
The matplotlib plot acts the same. If I try to use the zoom tool then it doesn't even display the crosshair. There may not be any way around this from the plotting side.

**Comment by nvaytet** [comment]
Maybe `matplotlib` is not really the right tool for this....

**Comment by matthew-d-jones** [comment]
Jupyter itself also uses the same event loop, so I'm starting to think the answer is that the data streaming consumption/aggregation should run on its own event loop in a separate process.
Either way tackling this problem is clearly not in the scope of this PR.

**Review by matthew-d-jones — APPROVED** [review]
I've tested different data types and plot dimensionality and found no issues other than those described in the comments when used in conjunction with `data_stream()`.
Tests and impl look good to me.

File: python/src/scipp/plotting/model.py [hunk 7]
```diff
@@ -86,14 +86,14 @@ def __init__(self,
...
-            self.data_arrays[name] = PlotArray(data=array.data,
-                                               meta=coord_list)
+            self.backup.update({name: {"array": array, "coords": coord_list}})
 
-            # Include masks
-            for m in array.masks:
-                self.data_arrays[name].masks[m] = array.masks[m]
+        # Save a copy of the name for simpler access
+        # Note this needs to be done before calling update_data_arrays
+        self.name = name
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/model.py:93]: The name of the last item?

File: python/src/scipp/plotting/model.py [hunk 8]
```diff
@@ -229,6 +227,13 @@ def _make_profile(self, profile, dim, mask_info):
         values["masks"] = self._make_masks(profile, mask_info=mask_info)
         return values
 
+    def update_data_arrays(self):
+        for name, item in self.backup.items():
+            self.data_arrays[name] = sc.DataArray(
+                data=item["array"].data,
+                coords=item["coords"],
+                masks=item["array"].masks.to_dict())
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/model.py:235]: Does not appear to use `self.name`, so I do not understand the comment above.

File: python/src/scipp/plotting/model2d.py [hunk 9]
```diff
@@ -147,8 +158,8 @@ def update_profile(self,
         # Find indices of pixel where cursor lies
         dimx = self.displayed_dims['x']
         dimy = self.displayed_dims['y']
-        x = self._model.data.meta[dimx]
-        y = self._model.data.meta[dimy]
+        x = self._model.data.coords[dimx]
+        y = self._model.data.coords[dimy]
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/model2d.py:162]: Why not `meta`?

File: python/src/scipp/plotting/objects.py [hunk 10]
```diff
@@ -6,7 +6,90 @@
...
+    def __init__(self, *args, **kwargs):
+        self._items = dict(*args, **kwargs)
+
+    def __getitem__(self, key):
+        return self._items[key]
+
+    def __setitem__(self, key, value):
+        self._items[key] = value
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/objects.py:25]: Do we need this? Should this be exposed to the user?

File: python/src/scipp/plotting/resampling_model.py [hunk 11]
```diff
@@ -85,7 +84,7 @@ def _call_resample(self):
         for dim, s in self.bounds.items():
             dim = str(dim)
             if s is None:
-                coord = self._array.meta[dim]
+                coord = self._array.coords[dim]
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:87]: Why not `meta` it was there for a reason?

File: python/src/scipp/plotting/resampling_model.py [hunk 12]
```diff
@@ -96,13 +95,17 @@ def _call_resample(self):
                 low, high = s
                 if isinstance(low, int):
                     out = out[dim, low:high]
-                    low = out.meta[dim][dim, 0]
-                    high = out.meta[dim][dim, -1]
+                    low = out.coords[dim][dim, 0]
+                    high = out.coords[dim][dim, -1]
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:99]: Why not `meta`?

File: python/src/scipp/plotting/resampling_model.py [hunk 13]
```diff
@@ -120,22 +123,32 @@ def _call_resample(self):
             self._home = self._resampled
             self._home_params = self._resampled_params
 
+    def update_array(self, array):
+        """
+        Update the internal array with a new array.
+        """
+        self._array = array
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:130]: Should maybe rename the `__init__` arg `data` to `array` for consistency? Also, maybe make `array` a property instead?

File: python/src/scipp/plotting/resampling_model.py [hunk 14]
```diff
@@ -120,22 +123,32 @@ def _call_resample(self):
...
+
+    def reset_params(self):
+        """
+        Reset view parameters to force a plot update.
+        """
+        self._home_params = None
+        self._resampled_params = None
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:137]: Is `reset()` a better name? The client does not really know or have to know that it is params that are being reset?

File: python/src/scipp/plotting/resampling_model.py [hunk 15]
```diff
@@ -120,22 +123,32 @@ def _call_resample(self):
...
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # TODO See #1469. This is a temporary hack to work around the
         # conversion of coords to edges in model.py.
-        new_meta = {name: var for name, var in self._array.meta.items()}
-        self._array = PlotArray(data=self._array.data,
-                                masks=self._array.masks,
-                                meta=new_meta)
-        for name, var in self._array.meta.items():
+        self._array = self._array.copy(deep=False)
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:145]: Something looks inconsistent here: `ResamplingModel.update_array` does not make a shallow copy, but here we do in the constructor?

File: python/src/scipp/plotting/resampling_model.py [hunk 16]
```diff
@@ -120,22 +123,32 @@ def _call_resample(self):
...
+        for name, var in self._array.coords.items():
             if len(var.dims) == 0:
                 continue
             dim = var.dims[-1]
-            if name not in self._array.data.bins.meta:
-                self._array.meta[name] = to_bin_centers(var, dim)
+            if name not in self._array.data.bins.coords:
+                self._array.coords[name] = to_bin_centers(var, dim)
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:151]: Shouldn't all this be called every time someone uses `update_array()`?

File: python/src/scipp/plotting/resampling_model.py [hunk 17]
```diff
@@ -147,14 +160,12 @@ def _resample(self, array):
             # Must specify bounds for final dim despite handling by `histogram`
             # below: If coord is ragged binning would throw otherwise.
             bounds = sc.concatenate(edges[dim, 0], edges[dim, -1], dim)
-            binned = sc.bin_with_coords(array.data, array.meta,
-                                        self.edges[:-1] + [bounds], [])
+            binned = sc.bin(array, self.edges[:-1] + [bounds], [], [])
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:163]: Use keyword args?

File: python/src/scipp/plotting/resampling_model.py [hunk 18]
```diff
@@ -147,14 +160,12 @@ def _resample(self, array):
...
             a = sc.histogram(binned, edges)
         else:
-            a = sc.bin_with_coords(array.data, array.meta, self.edges,
-                                   []).bins.sum()
+            a = sc.bin(array, self.edges, [], []).bins.sum()
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:166]: keyword args

File: python/src/scipp/plotting/resampling_model.py [hunk 19]
```diff
@@ -180,15 +191,16 @@ def __init__(self, array, **kwargs):
...
 
     def _to_density(self, array):
-        array = PlotArray(data=array.data.astype(sc.dtype.float64),
-                          meta=array.meta,
-                          masks=array.masks)
-        for dim in array.data.dims:
-            coord = array.meta[dim]
+        # If we want to be able to use redraw() on floating point data, we must
+        # avoid the copy where possible.
+        # if array.dtype not in [sc.dtype.float64, sc.dtype.float32]:
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:196]: commented code?

File: python/src/scipp/plotting/model3d.py [hunk 20]
```diff
@@ -121,7 +121,7 @@ def update_data(self, slices, mask_info):
                                  variances=np.zeros(shape),
                                  dtype=data_slice.data.dtype,
                                  unit=sc.units.one),
-                masks=data_slice.masks)
+                masks=data_slice.masks.to_dict())
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/model3d.py:124]: This now appears to be the only place where `to_dict` is used? Maybe we should avoid adding it to scipp objects everywhere (in `__init__.py` above) and just to this locally?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: python/src/scipp/plotting/model.py (hunk 7 above)
Comment (nvaytet) [reply python/src/scipp/plotting/model.py:93]: Yes, this has always been the case.
For 1d plots, there can be more than one item in in the dict, but `self.name` is not used.
For 2+d plots, there can only be one name.
I've always thought it was not a great way of setting the name, but I didn't find a better way.

File: python/src/scipp/plotting/model.py (hunk 8 above)
Comment (nvaytet) [reply python/src/scipp/plotting/model.py:235]: The overload of `update_data_arrays` in `model2d.py` does make use of `self.name`.

File: python/src/scipp/plotting/model2d.py (hunk 9 above)
Comment (nvaytet) [reply python/src/scipp/plotting/model2d.py:162]: Because I think by construction there should only be coords, no attrs, because we pick and choose only the coords we need when we first create the plot.
It used to be `meta` because `PlotArray` was looking at the original object, and this had to search both coords and attrs.
Now, we make a new DataArray with a new set of coords.

That said, I don't mind reverting to `meta` to minimize the number of changes.

Comment (SimonHeybrock) [reply python/src/scipp/plotting/model2d.py:162]: I'd suggest reverting to `meta`, given that it may not be unlikely that we may not mess with the coords in the future before this code is reached.

File: python/src/scipp/plotting/resampling_model.py (hunk 15 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:145]: I am making a shallow copy here because i am modifying the coords below, so i need a new dict in order not to overwrite the original coords with bin centers.

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:145]: I knw, my point is: Shouldn't you do the same when `update_array()` is called?

File: python/src/scipp/plotting/resampling_model.py (hunk 16 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:151]: It should if the coordinates are also modified.
Right now we are not supporting `redraw` for modified coords.
So I thought it would be cheaper to only do it once.
This might also disappear in the future when we remove the conversion to bin edges.

I could also just add a comment in the code explaining just this.

File: python/src/scipp/plotting/model3d.py (hunk 20 above)
Comment (nvaytet) [reply python/src/scipp/plotting/model3d.py:124]: hmm i found 3 places where it is used:
```
(base) nvaytet@this:~/code/scipp/source/python/src/scipp/plotting$ gr to_dict
__init__.py:    if hasattr(meta, "to_dict"):
__init__.py:        meta = meta.to_dict()
model3d.py:                masks=data_slice.masks.to_dict())
model.py:                masks=item["array"].masks.to_dict())
```

Comment (nvaytet) [reply python/src/scipp/plotting/model3d.py:124]: I'm thinking of maybe renaming this to `asdict` to be more like `astype`? Thoughts?

File: python/src/scipp/plotting/resampling_model.py (hunk 16 above)
Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:151]: Well, but aren't the shiny modifed coords gone once you call `update_array()`, since that overwrites `self._array`?

File: python/src/scipp/plotting/model3d.py (hunk 20 above)
Comment (SimonHeybrock) [reply python/src/scipp/plotting/model3d.py:124]: Make it a free function and keep it plotting internal?

File: python/src/scipp/plotting/resampling_model.py (hunk 15 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:145]: From Slack conversation:

Neil Vaytet: Can i ask you again why we needed to switch to bin centers here? for the binned data

Simon Heybrock: Because sc.bin distinguishes between grouping (as used for the spectrum numbers) and binning (as used for TOF).
Events to not store their spectrum number, so it does not support binning into "spectrum edges".
Basically, there is only a spectrum coord for the bins, not for the events.
and in that case only grouping is supported, not binning.

Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:145]: ok then I agree, i need to also do this when `update_array()` is called.

File: python/src/scipp/plotting/model3d.py (hunk 20 above)
Comment (nvaytet) [reply python/src/scipp/plotting/model3d.py:124]: In `plotting/tools.py` ?

Comment (SimonHeybrock) [reply python/src/scipp/plotting/model3d.py:124]: I don't care much, the point is that we have to be careful about adding new functions to the public API, since sooner or later these will become "hard" to remove or modify since many users will start depending on them.

File: python/src/scipp/plotting/resampling_model.py (hunk 13 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:130]: i would say change the argument to the init, but not sure about the property name, as we are not really returning the array, we are returning `self._resampled`

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:130]: Well, that is another property?

File: python/src/scipp/__init__.py [hunk 21]
```diff
@@ -25,7 +26,7 @@ def custom_formatwarning(msg, *args, **kwargs):
 from ._scipp import __version__
 # Import classes
 from ._scipp.core import Variable, DataArray, Dataset, GroupByDataArray, \
-                         GroupByDataset, Unit
+                         GroupByDataset, Unit, Coords, Masks
```
Comment (SimonHeybrock) [inline python/src/scipp/__init__.py:29]: Is this useful (or used)? If not revert?

Also wodnering why the `GroupBy*` classes are imported? If only for setting extra properties below...

File: python/src/scipp/plotting/resampling_model.py [hunk 22]
```diff
@@ -101,8 +100,12 @@ def _call_resample(self):
                 else:
                     # cannot pre-select bins for multi-dim coords
                     if len(out.meta[dim].dims) == 1:
-                        out = out[sc.get_slice_params(out.data, out.meta[dim],
-                                                      low, high)]
+                        # TODO: for now, slicing in scipp does not support step
+                        # and pybind11 py::slice requires a step
+                        # (which is always 1 here)
+                        sl = sc.get_slice_params(out.data, out.meta[dim], low,
+                                                 high)
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:107]: Would it be a better fix to just change `get_slice_params` to return with a step of 1 (to be honest I still don't understand why this change is needed now, was something broken?)?

File: python/src/scipp/plotting/resampling_model.py [hunk 23]
```diff
@@ -147,16 +165,18 @@ def _resample(self, array):
...
-            a = sc.bin_with_coords(array.data, array.meta, self.edges,
-                                   []).bins.sum()
+            a = sc.bin(array=array, edges=self.edges).bins.sum()
         for name, mask in array.masks.items():
             a.masks[name] = self._rebin(mask, array.meta)
         return a
 
+    def update_array(self, *args, **kwargs):
+        super().update_array(*args, **kwargs)
+        self._recenter_coords()
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:178]: Have you consider reverting the design, which might be slightly cleaner: Let the base-class `update_array()` call a method that can be overridden by child classes. Something like:

```python
# ResamplingModel
@array.setter
def array(self, array):
  self._array = array
  self._data_updated()

# child classes
def _data_updated(self):
   # whatever
```

File: python/src/scipp/__init__.py (hunk 21 above)
Comment (nvaytet) [reply python/src/scipp/__init__.py:29]: Yes, I was setting a method on the `Coords` and `Masks` before, but not anymore.
`GroupBy*` are imported only for setting the extra properties, yes.

File: python/src/scipp/plotting/resampling_model.py (hunk 22 above)
Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:107]: It because before we were going through the `_maybe_slice` mechanism where we were extracting the slice's `start` and `stop`: https://github.com/scipp/scipp/pull/1907/files#diff-f7dd39172a47ac585ef2a5c0714a8f5ba6acb848db46f59cdea519f4706a8091L73
(I think we were always in bin edges?)

The `py::slice` object always has a step, to it tries to slice the DataArray with `5:10:1` and we don't support the step of 1 in scipp, hence the TODO.

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:107]: I don't understand. We *only* support `step=1`.

Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:108]: Yes but you can't write `d['tof', 5:10:1]`, you have to write `d['tof', 5:10]`.

Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:108]: The first one raises an exception

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:108]: Hmm, works for variables. Can you just fix this, so it also works for data arrays?

Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:108]: ok, i didn't know it worked for Variables!

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:108]: The issue appears to be that this wrongly selects some value-based slicing overload, rather than positional.

File: python/src/scipp/plotting/resampling_model.py [hunk 24]
```diff
@@ -120,22 +120,37 @@ def _call_resample(self):
             self._home = self._resampled
             self._home_params = self._resampled_params
 
+    def _make_array(self, array):
+        return array
```
Comment (SimonHeybrock) [inline python/src/scipp/plotting/resampling_model.py:124]: So this is used by `ResamplingCountsModel` which does not override?

Comment (nvaytet) [reply python/src/scipp/plotting/resampling_model.py:124]: yes

Comment (SimonHeybrock) [reply python/src/scipp/plotting/resampling_model.py:124]: Maybe move it there then?

**Comment by nvaytet** [comment]
Do not merge, I found another issue with the units on the axes.

**Comment by nvaytet** [comment]
> Do not merge, I found another issue with the units on the axes.

Ok i found and fixed it.

File: python/src/scipp/_utils/to_string.py [hunk 25]
```diff
@@ -18,7 +18,7 @@ def name_with_unit(var=None, name=None, log=False):
         text = "log\u2081\u2080(" + text + ")"
     if var is not None:
         if var.bins is not None:
-            text += " [{}]".format(var.bins.data.unit)
+            text += " [{}]".format(var.bins.constituents["data"].unit)
```
Comment (SimonHeybrock) [inline python/src/scipp/_utils/to_string.py:21]: Note there is now `var.events.unit` which maps to the same.

**Review by SimonHeybrock — APPROVED** [review]
